File: problems/0002-add-two-numbers/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addDigits(l1 *ListNode, l2 *ListNode, carry int) *ListNode {
	if l1 == nil && l2 == nil {
		panic("Invalid state, both lists are nil")
	}
	l1Val := 0
	l2Val := 0

	if l1 != nil {
		l1Val = l1.Val
	}
	if l2 != nil {
		l2Val = l2.Val
	}
	value := carry + l1Val + l2Val
	digitCarry := value / 10
	digitValue := value % 10
	hasNextL1 := false
	hasNextL2 := false
	if l1 != nil && l1.Next != nil {
		hasNextL1 = true
	}
	if l2 != nil && l2.Next != nil {
		hasNextL2 = true
	}
	if !hasNextL1 && !hasNextL2 {
		var carryNode *ListNode
		if digitCarry > 0 {
			carryNode = &ListNode{Val: digitCarry}
		}
		return &ListNode{Val: digitValue, Next: carryNode}
	}
	var l1Next *ListNode
	if l1 != nil {
		l1Next = l1.Next
	}
	var l2Next *ListNode
	if l2 != nil {
		l2Next = l2.Next
	}
	return &ListNode{
		Val:  digitValue,
		Next: addDigits(l1Next, l2Next, digitCarry),
	}
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil || l2 == nil {
		return nil
	}
	if l1 == nil {
		return l2
	}
	if l2 == nil {
		return l1
	}
	return addDigits(l1, l2, 0)
}

File: problems/0002-add-two-numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode addDigits(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null) {
            throw new RuntimeException("Invalid state, both lists are null");
        }
        int value = carry + (l1 == null ? 0 : l1.val) + (l2 == null ? 0 : l2.val);
        int digitCarry = value / 10;
        int digitValue = value % 10;
        boolean hasNextL1 = l1 != null && l1.next != null;
        boolean hasNextL2 = l2 != null && l2.next != null;
        if (!hasNextL1 && !hasNextL2) {
            ListNode carryNode = digitCarry > 0 ? new ListNode(digitCarry) : null;
            return new ListNode(digitValue, carryNode);
        }
        return new ListNode(
            digitValue,
            addDigits(
                l1 != null ? l1.next : null,
                l2 != null ? l2.next : null,
                digitCarry
            )
        );
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null) {
            return null;
        }
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }

        return this.addDigits(l1, l2, 0);
    }
}